### PR TITLE
C++: Fix some FPs in cpp/missing-check-scanf (third attempt!)

### PR DESCRIPTION
**Pull request by MathiasVP**
This PR brings back the changes from https://github.com/github/codeql/pull/17938 and fixes the performance by inlining the `ensures` (and to some degree the `compares`) predicates which became too large after the improvements.

Commit-by-commit review recommended:

- 20dfbdc5cc4caa1974e53fa00639872371e53e5d reverts the revert
- 6f73aa552dfb5d3156f5c237c5aa10a700f8cff8 does a small refactoring which I've been meaning to do for a while now that we can use `abstract` classes in user-facing classes by using final extensions.
- 404dd334984a2933116178b5dc76c5527ccd66aa and 3bdfdd0573ebb99f3a1a84f77df0ee66539da5e7 do the actual performance fix: we inline the `ensures` predicates, and move the `cache`ing to the implementation predicates. This should that we don't get expensive recomputations _and_ that the user-facing predicates are never materialized.

As an extra bonus, the inlining makes the performance fix added in 19d53fba8cf1d6ab68bd37a81d5acd6f1f46535b unnecessary 🎉

After several rounds of join order fixes QA looks good 🎉 (0.55% analysis time difference). You can find the QA run in the backlink if you're curious.

### CODE REVIEW

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 1]
```diff
@@ -727,630 +714,730 @@
...
+  }
+
+  /*
+   * Simplification of inequality expressions
+   * Simplify conditions in the source to the canonical form l < r + k.
+   */
+
+  /** Holds if `left < right + k` evaluates to `isLt` given that test is `testIsTrue`. */
+  cached
+  predicate compares_lt(
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:1098]: Fixed in 913357b70d8d531b337860ca9568d9c2d49061f7

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 2]
```diff
@@ -727,630 +714,730 @@
...
-    compares_lt(cmp, left, rhs.getAUse(), c, isLt, value) and
-    (
-      right = rhs.getLeftOperand() and x = int_value(rhs.getRight())
+  /**
+   * Holds if `left == right + k` is `areEqual` given that test is `testIsTrue`.
+   *
+   * Beware making mistaken logical implications here relating `areEqual` and `testIsTrue`.
+   */
+  cached
+  predicate compares_eq(
```
[the remark this answers is not in the document]
Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:824]: Fixed in 913357b70d8d531b337860ca9568d9c2d49061f7

File: cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll [hunk 3]
```diff
@@ -142,28 +142,15 @@ private class LogicalNotValueNumber extends ValueNumber {
...
  */
-cached
-class GuardCondition extends Expr {
-  cached
-  GuardCondition() {
-    exists(IRGuardCondition ir | this = ir.getUnconvertedResultExpression())
-    or
-    // no binary operators in the IR
-    this.(BinaryLogicalOperation).getAnOperand() instanceof GuardCondition
-  }
```
Comment (jketema) [inline cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:72]: Why was this removed?

Comment (MathiasVP) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:72]: Before the introduction of `final` aliases this class couldn't be made `abstract` (since it would expose a abstract class from our libraries to queries). So instead of having this class be abstract we followed the pattern of:
1. Specifying each subclass of this "pseudo-abstract" class in the charpred
2. Provide a subclass that extended this "pseudo-abstract" class and repeated the required charpred to define the subclass.

That is, we've had to do:
```qll
class Base {
  Base() {
     myImpl1(this) or myImpl2(this)
  }
}

class MyImpl1 extends Base {
  MyImpl1() {
    myImpl1(this)
  }
}

class MyImpl2 extends Base {
  MyImpl2() {
    myImpl1(this)
  }
}
```

Now that we have final classes we can get rid of this pattern (as we've done many other places) and actually use abstract classes. And abstract classes don't need this charpred since they're defined as the union of all the subclasses.

That is, we can rewrite the above example to:
```qll
private abstract class BaseImpl { }

final class Base = BaseImpl;

class MyImpl1 extends BaseImpl {
  MyImpl1() {
    myImpl1(this)
  }
}

class MyImpl2 extends BaseImpl {
  MyImpl2() {
    myImpl1(this)
  }
}
```

Does that make sense?

Comment (jketema) [reply cpp/ql/lib/semmle/code/cpp/controlflow/IRGuards.qll:72]: Makes total sense.

**Review by jketema — APPROVED** [review]
Additional changes after the revert-revert make sense.

**Comment by MathiasVP** [comment]
Thanks a lot, Jeroen! DCA results are in and shows nothing of interest
